### PR TITLE
Fix github issue #205

### PR DESCRIPTION
**Pull request by takikawa**
This PR fixes issue #205 and some related bugs I found, but unfortunately there is one test that doesn't pass right now.

I think that the test actually exhibits broken behavior, so I wanted a second opinion on what to do about it.

The problematic test is this one: https://github.com/racket/typed-racket/blob/master/typed-racket-test/succeed/poly-struct-parent.rkt

On my PR branch, line 25 fails to typecheck because the predicate on that line has a filter that overlaps with the `Box2` type. Unfortunately, the original behavior (that the types don't overlap) is also wrong because it makes that whole conditional branch into dead code (try adding `(+ 1 "foo")` in that cond branch). Anyone have suggestions on how to fix this test?

### CODE REVIEW

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 1]
```diff
@@ -282,15 +295,28 @@
...
+             ;; no parent, so trivially true
+             ['() #t]))
+         (define desc
+           (struct-desc parent-fields types tvars mutable parent-mutable #f))
          (parsed-struct (make-Prefab key (append parent-fields types))
                         names desc (struct-info-property nm/par) #f)]
         [else
+         (define parent-mutable
+           ;; Only valid as long as typed structs must be
+           ;; either fully mutable or fully immutable
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt]: Can we prepare for what happens when we fix this?

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 2]
```diff
@@ -311,7 +337,8 @@
   (define parent-tys (map fld-t (get-flds parent-type)))
 
   (define names (get-struct-names nm fld-names #f))
-  (define desc (struct-desc parent-tys tys null #t #f))
+  ;; built-in structs are assumed to be immutable with immutable parents
+  (define desc (struct-desc parent-tys tys null #f #f #f))
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt]: Is this true everywhere?

Comment (takikawa) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt]: Yes for the current set of built-in structs, no if that ever changes. This is certainly more correct than the status quo though.

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt (hunk 1 above)
Comment (takikawa) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt]: That would involve changing a lot more code that depends on this fact. I think it'd be better put in a different PR (and really I think we should just fix it in the next TR or something).

**Comment by takikawa** [comment]
Any ideas on what to do about that failing test case?

**Comment by samth** [comment]
I don't totally understand what you're saying about the test case.

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt (hunk 2 above)
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt]: Ok, I'll assume that the built-in structs aren't going to grow some mutable data.

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt (hunk 1 above)
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt]: Right, I just meant can we avoid adding more code that's obviously unsound when that happens.

And I agree we should fix that soon.

**Comment by takikawa** [comment]
@samth I was saying that the original test case is wrong, in that it only succeeds to typecheck because a branch is declared "dead" but it really isn't. The PR causes the branch to be live again, but then the RHS doesn't typecheck. I wasn't sure what the best way to fix the test is though.

**Comment by samth** [comment]
Right, but why doesn't `b` have the appropriate type there?

**Comment by takikawa** [comment]
Because the filter on the predicate `Strict-Box2` specifies a `(Strict-Box2 Any)` and therefore that's the type of `b`. This isn't compatible with the return type, which is `(Box2 A)`.

**Comment by samth** [comment]
Shouldn't we refine that to `(Strict-Box2 A)` and then have the test succeed?

**Comment by takikawa** [comment]
I don't think we want to build in the logic that `(Foo Any)` intersected with `(Parent-Foo X)` is `(Foo X)` though because that's not valid in general. Or is that not what you're suggesting?

**Comment by samth** [comment]
That is what I'm suggesting, but I think it is valid.

**Comment by takikawa** [comment]
Well, I can't actually think of a counterexample so you may be right. I'll see if I can implement that and see if it works out.

**Comment by samth** [comment]
Status?

**Comment by takikawa** [comment]
Blocked on implementing that intersection. Unlikely to do it any time soon.

**Comment by samth** [comment]
Ok. I'd like to get #205 fixed, since it's a soundness issue, so is there something simpler we can do? Alternatively, I can look at `restrict`.

**Comment by takikawa** [comment]
Merged because the one remaining issue with this PR went away, possibly due to recent intersection type additions.
